pkg/app/piped/appconfigreporter/appconfigreporter.go

```diff
@@ -0,0 +1,337 @@
+	}
+
+	if err := r.updateUnregisteredApps(ctx, registeredAppPaths); err != nil {
+		return fmt.Errorf("failed to update unregistered applications: %w", err)
+	}
+	if err := r.updateRegisteredApps(ctx, registeredAppPaths); err != nil {
+		return fmt.Errorf("failed to update registered applications: %w", err)
+	}
+
+	return nil
+}
+
+// updateUnregisteredApps sends all unregistered application configurations to the control-plane.
+func (r *Reporter) updateUnregisteredApps(ctx context.Context, registeredAppPaths map[string]struct{}) error {
+	apps := make([]*model.ApplicationInfo, 0)
+	for repoID, repo := range r.gitRepos {
+		as, err := r.findUnregisteredApps(ctx, repo.GetPath(), repoID, registeredAppPaths)
+		if err != nil {
+			return err
+		}
+		apps = append(apps, as...)
+	}
+	if len(apps) == 0 {
+		return nil
+	}
+
+	_, err := r.apiClient.ReportUnregisteredApplicationConfigurations(
+		ctx,
+		&pipedservice.ReportUnregisteredApplicationConfigurationsRequest{
+			Applications: apps,
+		},
+	)
+	if err != nil {
+		return fmt.Errorf("failed to put the latest unregistered application configurations: %w", err)
+	}
+	return nil
+}
+
+// findUnregisteredApps finds out unregistered application info in the given git repository.
+func (r *Reporter) findUnregisteredApps(ctx context.Context, repoPath, repoID string, registeredAppPaths map[string]struct{}) ([]*model.ApplicationInfo, error) {
+	apps := make([]*model.ApplicationInfo, 0)
+	err := fs.WalkDir(r.fsys, strings.TrimPrefix(repoPath, "/"), func(path string, d fs.DirEntry, err error) error {
+		if err != nil {
+			return err
+		}
+		if d.IsDir() {
+			return nil
+		}
+
+		if shouldSkip(repoID, repoPath, filepath.Base(path), registeredAppPaths, false) {
+			return nil
+		}
+
+		appInfo, err := r.readApplicationInfo(ctx, repoPath, path)
+		if err != nil {
+			r.logger.Error("failed to read application info",
+				zap.String("repo-id", repoID),
+				zap.String("config-file-path", path),
+				zap.Error(err),
+			)
+			return nil
+		}
+		apps = append(apps, appInfo)
+		return nil
+	})
+	if err != nil {
+		return nil, fmt.Errorf("failed to inspect files under %s: %w", repoPath, err)
+	}
+	return apps, nil
+}
+
+// updateRegisteredApps sends application configurations that have changed since the last time to the control-plane.
+func (r *Reporter) updateRegisteredApps(ctx context.Context, registeredAppPaths map[string]struct{}) (err error) {
+	apps := make([]*model.ApplicationInfo, 0)
+	for repoID, repo := range r.gitRepos {
+		var headCommit git.Commit
+		headCommit, err = repo.GetLatestCommit(ctx)
+		if err != nil {
+			return fmt.Errorf("failed to get the latest commit of %s: %w", repoID, err)
+		}
+		var as []*model.ApplicationInfo
+		as, err = r.findRegisteredApps(ctx, repoID, repo, headCommit.Hash, registeredAppPaths)
+		if err != nil {
+			return err
+		}
+		apps = append(apps, as...)
+		id := repoID
+		defer func() {
+			if err == nil {
+				r.lastScannedCommits[id] = headCommit.Hash
+			}
+		}()
```

Comment:
I think to be simple we should avoid `defer` in the loop.
So instead of that, we can mark the commit list in a variable and save them in a single `defer`.

Reply:
Cool. It will reduce the amount of code.